### PR TITLE
Added basic interfaces and extended AbstractDiscoveryStrategy

### PR DESCRIPTION
**Pull request by noctarius**
Extended the DiscoveryStrategy interface and AbstractDiscoveryStrategy implementation to support discovery of a local's node metadata from the given runtime environment (like AWS, GCE, ...).

### CODE REVIEW

**Comment by noctarius** [comment]
I'm getting this error
```
23:55:13 [INFO] Starting audit...
23:55:13 /appdisk/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/main/java/com/hazelcast/transaction/impl/xa/XATransaction.java:157:33: warning: Unable to get class information for TransactionException.
23:55:13 /appdisk/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/main/java/com/hazelcast/transaction/impl/TransactionImpl.java:264:33: warning: Unable to get class information for TransactionException.
23:55:13 /appdisk/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/main/java/com/hazelcast/transaction/impl/Transaction.java:28:26: warning: Unable to get class information for TransactionException.
23:55:13 /appdisk/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/main/java/com/hazelcast/instance/Node.java:85: Line is longer than 130 characters (found 157).
23:55:13 /appdisk/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/main/java/com/hazelcast/query/IndexAwarePredicate.java:26: Type Javadoc comment is missing an @param <K> tag.
23:55:13 /appdisk/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/main/java/com/hazelcast/query/IndexAwarePredicate.java:26: Type Javadoc comment is missing an @param <V> tag.
```

Is there any issue in master as those classes are not changed?

**Comment by Donnerbart** [comment]
Please rebase, I fixed that error provisional.

**Comment by noctarius** [comment]
verify

**Comment by noctarius** [comment]
verify

**Comment by kobalski** [comment]
:+1: 
We also need to update plugin repos below which will be effected SPI changes.
https://github.com/hazelcast/hazelcast-zookeeper
https://github.com/hazelcast/hazelcast-azure
https://github.com/hazelcast/hazelcast-jclouds

File: hazelcast/src/main/java/com/hazelcast/spi/partitiongroup/PartitionGroupStrategy.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * to still provide all data. Similar is true for zones or physical hosts.</p>
+ * <p>Custom implementations of the PartitionGroupStrategy may add specific
+ * or additional behavior based on the provided environment and can
+ * be injected into Hazelcast by overriding
+ * {@link AbstractDiscoveryStrategy#getPartitionGroupStrategy()}.</p>
+ */
+public interface PartitionGroupStrategy {
+
+    /**
+     * TODO Has probably to be changed :)
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/partitiongroup/PartitionGroupStrategy.java]: when? :)

**Comment by bilalyasar** [comment]
can you add some tests?

File: hazelcast/src/main/java/com/hazelcast/spi/partitiongroup/PartitionGroupStrategy.java (hunk 1 above)
Comment (kobalski) [inline hazelcast/src/main/java/com/hazelcast/spi/partitiongroup/PartitionGroupStrategy.java]: There will be a follow up pr which implements partition groups over aws zones.

**Comment by noctarius** [comment]
There can't be a test yet, as the actual code to use those injected attributes is basically missing or do you look for a test to just see the attributes are injected into the member after creation? Yeah can do that :)

**Comment by noctarius** [comment]
run-lab-run

**Comment by noctarius** [comment]
@kobalski If the plugins extend the AbstractDiscoveryStrategy they should still work anyhow you're right they should provide the metadata whenever possible :D

**Comment by noctarius** [comment]
run-lab-run

**Comment by mesutcelik** [comment]
👍

**Comment by pveentjer** [comment]
Since there are 2 thumbs up, this PR can be merged.

**Pull request merged**